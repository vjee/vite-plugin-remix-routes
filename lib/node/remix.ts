import { defineConventionalRoutes } from "@remix-run/dev/dist/config/routesConvention";
import {
  ConfigRoute,
  RouteManifest,
  defineRoutes,
} from "@remix-run/dev/dist/config/routes";
import type { PluginOptions } from "./index";
import type { AppConfig } from "@remix-run/dev/dist/config";
import type { RequireOnly } from "./utils";

export type RemixOptions = Pick<
  AppConfig,
  "appDirectory" | "routes" | "ignoredRouteFiles"
>;

type GetRouteOptions = Omit<PluginOptions, "importModule"> &
  RequireOnly<RemixOptions, "appDirectory">;

/**
 * See `readConfig` in @remix-run/dev/config.ts
 */
export async function getRoutes(options: GetRouteOptions) {
  const {
    appDirectory,
    is404Route = (route) => route.id.endsWith("/404"),
    ignoredRouteFiles,
    routes,
  } = options;

  const routeManifest: RouteManifest = {
    root: { path: "", id: "root", file: "routes/index" },
  };

  const conventionalRoutes = defineConventionalRoutes(
    appDirectory,
    ignoredRouteFiles
  );

  for (const key of Object.keys(conventionalRoutes)) {
    const route = conventionalRoutes[key];
    routeManifest[route.id] = {
      ...route,
      parentId: route.parentId || "root",
    };
  }

  if (routes) {
    let manualRoutes = await routes(defineRoutes);
    for (const key of Object.keys(manualRoutes)) {
      const route = manualRoutes[key];
      routeManifest[route.id] = {
        ...route,
        parentId: route.parentId || "root",
      };
    }
  }

  const routeConfig = createRoutes(routeManifest)[0].children;

  // This is not part of remix.
  const modifyRoute = (route: Route): Route => ({
    ...route,
    path: is404Route(route) ? "*" : route.path,
    children: route.children.map(modifyRoute),
  });

  return routeConfig.map(modifyRoute);
}

/**
 * See `createClientRoutes` in @remix-run/react/routes.tsx
 */
export function createRoutes(
  routeManifest: RouteManifest,
  parentId?: string
): Route[] {
  return Object.keys(routeManifest)
    .filter((key) => routeManifest[key].parentId === parentId)
    .map((key) => {
      const route = createRoute(routeManifest[key]);
      route.children = createRoutes(routeManifest, route.id);
      return route;
    });
}

/**
 * See `createClientRoute` in @remix-run/react/routes.tsx
 */
export function createRoute(route: ConfigRoute): Route {
  return {
    id: route.id,
    file: route.file,

    path: route.path || "",
    index: !!route.index,
    children: [],
  };
}

export interface Route {
  // custom properties
  id: string;
  file: string;

  // react-router route properties
  path: string;
  index: boolean;
  children: Route[];
}
