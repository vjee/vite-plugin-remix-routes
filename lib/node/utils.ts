import type { Route } from "./remix";

export type RequireOnly<Object, Keys extends keyof Object> = Omit<
  Object,
  Keys
> &
  Required<Pick<Object, Keys>>;

export interface Context {
  prefix: string;
  dataRouterCompatible?: boolean;
  importMode?: (route: Route) => "sync" | "async";
}

interface Components {
  sync: string[];
  async: string[];
}

export function stringifyRoutes(routes: Route[], context: Context) {
  const components: Components = { sync: [], async: [] };
  const routesString = routesToString(routes, context, components);

  return {
    routesString,
    componentsString: [...components.sync, ...components.async].join("\n"),
  };
}

function routesToString(
  routes: Route[],
  context: Context,
  components: Components
) {
  return (
    "[" +
    routes.map((route) => routeToString(route, context, components)).join(",") +
    "]"
  );
}

function routeToString(
  route: Route,
  context: Context,
  components: Components
): string {
  const componentName = getRouteComponentName(route);
  const componentPath = `${context.prefix}/${route.file}`;
  const importMode = context.importMode?.(route) || "sync";

  const props = new Map<string, string>();

  if (route.path !== "") {
    props.set("path", `'${route.path}'`);
  }

  if (context.dataRouterCompatible) {
    components.sync.push(
      `import * as ${componentName} from '${componentPath}';`
    );

    props.set(
      "element",
      `${componentName}.default ? createElement(${componentName}.default) : undefined`
    );
    props.set("loader", `${componentName}.loader`);
    props.set("action", `${componentName}.action`);
    props.set(
      "errorElement",
      `${componentName}.ErrorBoundary ? createElement(${componentName}.ErrorBoundary) : undefined`
    );
    props.set("handle", `${componentName}.handle`);
    props.set("shouldRevalidate", `${componentName}.shouldRevalidate`);
  } else {
    if (importMode === "async") {
      components.async.push(
        `const ${componentName} = () => import('${componentPath}');`
      );

      props.set("element", `createElement(lazy(${componentName}))`);
      props.set("importPromise", componentName);
    }

    if (importMode === "sync") {
      components.sync.push(`import ${componentName} from '${componentPath}';`);

      props.set("element", `createElement(${componentName})`);
    }
  }

  if (route.index === true) {
    props.set("index", "true");
  }

  if (route.children.length) {
    const children = routesToString(route.children, context, components);
    props.set("children", children);
  }

  return (
    "{" + [...props.entries()].map(([k, v]) => `${k}:${v}`).join(",") + "}"
  );
}

function getRouteComponentName(route: Route) {
  return route.id
    .split(/[/.]/)
    .map((str) => str.replace(/^\w/, (c) => c.toUpperCase()))
    .join("");
}
