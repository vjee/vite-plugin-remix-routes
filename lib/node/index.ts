import path from "node:path";
import fs from "node:fs";

import type { Plugin } from "vite";

import { getRoutes, RemixOptions } from "./remix";
import type { Route } from "./remix";
import { stringifyRoutes } from "./utils";

export interface Options extends PluginOptions, RemixOptions {}
export interface PluginOptions {
  /**
   * An absolute path to the folder containing the `routes` folder.
   * This will most likely be your `/src` folder.
   *
   * @deprecated Use `appDirectory` instead
   */
  appDir?: string;

  /**
   * A function that receives a `Route` to determine if the route's component
   * should be imported synchronously or asynchronously.
   *
   * @default () => "sync"
   */
  importMode?: (route: Route) => "async" | "sync";

  /**
   * A function that receives a `Route` to determine if it should be a 404 route. (`path="*"`)
   * By default this matches the same 404 file as Remix does.
   * Keep in mind this only receives top level routes, so you can't mark nested routes
   * as 404 routes.
   *
   * @default (route) => route.id === "routes/404"
   */
  is404Route?: (route: Route) => boolean;
}

function plugin(options: Options = {}): Plugin {
  const virtualModuleId = "virtual:remix-routes";

  const {
    appDir: _appDir,
    appDirectory = _appDir,
    importMode,
    ...otherOptions
  } = options;

  let dir = "";

  // Check both src for backwards compatibility and app for remix's normal default
  if (appDirectory) {
    dir = path.resolve(process.cwd(), appDirectory);
  } else {
    const srcDir = path.resolve(process.cwd(), "src");
    const appDir = path.resolve(process.cwd(), "app");

    if (!dir && fs.existsSync(srcDir)) {
      dir = srcDir;
    }

    if (!dir && fs.existsSync(appDir)) {
      dir = appDir;
    }

    if (!dir) {
      throw new Error(
        "[vite-plugin-remix-routes] appDirectory not found, please specify it in the plugin's config"
      );
    }
  }

  const prefix = `.${path.sep}${path.relative(process.cwd(), dir)}`;

  return {
    name: "vite-plugin-remix-routes",

    resolveId(id) {
      if (id === virtualModuleId) {
        return id;
      }
    },

    async load(id) {
      if (id === virtualModuleId) {
        const routes = await getRoutes({ appDirectory: dir, ...otherOptions });

        const { routesString, componentsString } = stringifyRoutes(routes, {
          prefix,
          importMode,
        });

        return (
          `import { createElement, lazy, useEffect } from 'react';\n` +
          `${componentsString}\n` +
          `export default ${routesString};\n`
        );
      }
    },
  };
}

export default plugin;
export { getRoutes, stringifyRoutes };

export type { Route };
