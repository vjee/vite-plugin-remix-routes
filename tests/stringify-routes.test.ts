import path from "node:path";
import url from "node:url";

import { getRoutes, stringifyRoutes } from "../lib/node";
import type { Context } from "../lib/node/utils";

const dirname = path.dirname(url.fileURLToPath(import.meta.url));
const appDirectory = path.join(dirname, "../examples/basic/app");

test("stringifyRoutes", async () => {
  const routes = await getRoutes({ appDirectory });
  const prefix = "/app";
  const importMode: Context["importMode"] = (route) =>
    route.id.startsWith("routes/demos/about") ? "async" : "sync";

  const { routesString, componentsString } = stringifyRoutes(routes, {
    prefix,
    importMode,
  });

  expect(routesString).toMatchSnapshot();
  expect(componentsString).toMatchSnapshot();
});

test("stringifyRoutes for dataBrowser", async () => {
  const routes = await getRoutes({ appDirectory });
  const prefix = "/app";
  const importMode: Context["importMode"] = () => "dataRoute";

  const { routesString, componentsString } = stringifyRoutes(routes, {
    prefix,
    importMode,
  });

  expect(routesString).toMatchSnapshot();
  expect(componentsString).toMatchSnapshot();
});
