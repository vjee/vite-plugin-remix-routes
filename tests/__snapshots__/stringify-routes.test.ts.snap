// Vitest Snapshot v1

exports[`stringifyRoutes 1`] = `"[{path:'one/two/three',element:createElement(RoutesOneTwoThree)},{path:'demos/params',element:createElement(RoutesDemosParams),children:[{element:createElement(RoutesDemosParamsIndex),index:true},{path:':id',element:createElement(RoutesDemosParams$id)}]},{path:'demos/about',element:createElement(lazy(RoutesDemosAbout)),loader:RoutesDemosAbout,children:[{element:createElement(lazy(RoutesDemosAboutIndex)),loader:RoutesDemosAboutIndex,index:true},{path:'whoa',element:createElement(lazy(RoutesDemosAboutWhoa)),loader:RoutesDemosAboutWhoa},{path:'*',element:createElement(lazy(RoutesDemosAbout404)),loader:RoutesDemosAbout404}]},{element:createElement(RoutesIndex),index:true},{path:'*',element:createElement(Routes404)}]"`;

exports[`stringifyRoutes 2`] = `
"import RoutesOneTwoThree from '/app/routes/one.two.three.tsx';
import RoutesDemosParams from '/app/routes/demos/params.tsx';
import RoutesDemosParamsIndex from '/app/routes/demos/params/index.tsx';
import RoutesDemosParams$id from '/app/routes/demos/params/$id.tsx';
import RoutesIndex from '/app/routes/index.tsx';
import Routes404 from '/app/routes/404.tsx';
const RoutesDemosAbout = () => import('/app/routes/demos/about.tsx');
const RoutesDemosAboutIndex = () => import('/app/routes/demos/about/index.tsx');
const RoutesDemosAboutWhoa = () => import('/app/routes/demos/about/whoa.tsx');
const RoutesDemosAbout404 = () => import('/app/routes/demos/about/404.tsx');"
`;

exports[`stringifyRoutes for dataBrowser 1`] = `"[{path:'one/two/three',element:createElement(RoutesOneTwoThree.default),loader:RoutesOneTwoThree.loader,action:RoutesOneTwoThree.action,errorElement:RoutesOneTwoThree.errorBoundary,handle:RoutesOneTwoThree.handle,shouldRevalidate:RoutesOneTwoThree.shouldRevalidate},{path:'demos/params',element:createElement(RoutesDemosParams.default),loader:RoutesDemosParams.loader,action:RoutesDemosParams.action,errorElement:RoutesDemosParams.errorBoundary,handle:RoutesDemosParams.handle,shouldRevalidate:RoutesDemosParams.shouldRevalidate,children:[{element:createElement(RoutesDemosParamsIndex.default),loader:RoutesDemosParamsIndex.loader,action:RoutesDemosParamsIndex.action,errorElement:RoutesDemosParamsIndex.errorBoundary,handle:RoutesDemosParamsIndex.handle,shouldRevalidate:RoutesDemosParamsIndex.shouldRevalidate,index:true},{path:':id',element:createElement(RoutesDemosParams$id.default),loader:RoutesDemosParams$id.loader,action:RoutesDemosParams$id.action,errorElement:RoutesDemosParams$id.errorBoundary,handle:RoutesDemosParams$id.handle,shouldRevalidate:RoutesDemosParams$id.shouldRevalidate}]},{path:'demos/about',element:createElement(RoutesDemosAbout.default),loader:RoutesDemosAbout.loader,action:RoutesDemosAbout.action,errorElement:RoutesDemosAbout.errorBoundary,handle:RoutesDemosAbout.handle,shouldRevalidate:RoutesDemosAbout.shouldRevalidate,children:[{element:createElement(RoutesDemosAboutIndex.default),loader:RoutesDemosAboutIndex.loader,action:RoutesDemosAboutIndex.action,errorElement:RoutesDemosAboutIndex.errorBoundary,handle:RoutesDemosAboutIndex.handle,shouldRevalidate:RoutesDemosAboutIndex.shouldRevalidate,index:true},{path:'whoa',element:createElement(RoutesDemosAboutWhoa.default),loader:RoutesDemosAboutWhoa.loader,action:RoutesDemosAboutWhoa.action,errorElement:RoutesDemosAboutWhoa.errorBoundary,handle:RoutesDemosAboutWhoa.handle,shouldRevalidate:RoutesDemosAboutWhoa.shouldRevalidate},{path:'*',element:createElement(RoutesDemosAbout404.default),loader:RoutesDemosAbout404.loader,action:RoutesDemosAbout404.action,errorElement:RoutesDemosAbout404.errorBoundary,handle:RoutesDemosAbout404.handle,shouldRevalidate:RoutesDemosAbout404.shouldRevalidate}]},{element:createElement(RoutesIndex.default),loader:RoutesIndex.loader,action:RoutesIndex.action,errorElement:RoutesIndex.errorBoundary,handle:RoutesIndex.handle,shouldRevalidate:RoutesIndex.shouldRevalidate,index:true},{path:'*',element:createElement(Routes404.default),loader:Routes404.loader,action:Routes404.action,errorElement:Routes404.errorBoundary,handle:Routes404.handle,shouldRevalidate:Routes404.shouldRevalidate}]"`;

exports[`stringifyRoutes for dataBrowser 2`] = `
"import * as RoutesOneTwoThree from '/app/routes/one.two.three.tsx';
import * as RoutesDemosParams from '/app/routes/demos/params.tsx';
import * as RoutesDemosParamsIndex from '/app/routes/demos/params/index.tsx';
import * as RoutesDemosParams$id from '/app/routes/demos/params/$id.tsx';
import * as RoutesDemosAbout from '/app/routes/demos/about.tsx';
import * as RoutesDemosAboutIndex from '/app/routes/demos/about/index.tsx';
import * as RoutesDemosAboutWhoa from '/app/routes/demos/about/whoa.tsx';
import * as RoutesDemosAbout404 from '/app/routes/demos/about/404.tsx';
import * as RoutesIndex from '/app/routes/index.tsx';
import * as Routes404 from '/app/routes/404.tsx';"
`;
